extensions/druid-rocketmq/src/main/java/io/druid/firehose/rocketmq/RocketMQFirehoseFactory.java

```diff
@@ -0,0 +1,537 @@
+import com.alibaba.rocketmq.client.consumer.MessageQueueListener;
+import com.alibaba.rocketmq.client.consumer.PullResult;
+import com.alibaba.rocketmq.client.consumer.store.OffsetStore;
+import com.alibaba.rocketmq.client.consumer.store.ReadOffsetType;
+import com.alibaba.rocketmq.client.exception.MQBrokerException;
+import com.alibaba.rocketmq.client.exception.MQClientException;
+import com.alibaba.rocketmq.common.ServiceThread;
+import com.alibaba.rocketmq.common.message.MessageExt;
+import com.alibaba.rocketmq.common.message.MessageQueue;
+import com.alibaba.rocketmq.common.protocol.heartbeat.MessageModel;
+import com.alibaba.rocketmq.remoting.exception.RemotingException;
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.common.collect.Sets;
+import com.metamx.common.parsers.ParseException;
+import io.druid.data.input.ByteBufferInputRowParser;
+import io.druid.data.input.Firehose;
+import io.druid.data.input.FirehoseFactory;
+import io.druid.data.input.InputRow;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.*;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentSkipListSet;
+import java.util.concurrent.CountDownLatch;
+
+public class RocketMQFirehoseFactory implements FirehoseFactory<ByteBufferInputRowParser> {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(RocketMQFirehoseFactory.class);
+
+    /**
+     * Passed in configuration for consumer client.
+     */
+    @JsonProperty
+    private final Properties consumerProps;
+
+    /**
+     * Consumer group.
+     */
+    @JsonProperty
+    private final String consumerGroup;
+
+    /**
+     * Topics to consume.
+     * Multiple topics are separated by comma ",".
+     */
+    @JsonProperty
+    private final String feed;
+
+
+    /**
+     * Store messages that are fetched from brokers but not yet delivered to druid via fire hose.
+     */
+    private ConcurrentHashMap<MessageQueue, ConcurrentSkipListSet<MessageExt>> messageQueueTreeSetMap = new ConcurrentHashMap<>();
+
+    /**
+     * Store message consuming status.
+     */
+    private ConcurrentHashMap<MessageQueue, ConcurrentSkipListSet<Long>> windows = new ConcurrentHashMap<>();
+
+    /**
+     * Default pull batch size.
+     */
+    private static final int PULL_BATCH_SIZE = 32;
+
+    @JsonCreator
+    public RocketMQFirehoseFactory(@JsonProperty("consumerProps") Properties consumerProps,
+                                   @JsonProperty("consumerGroup") String consumerGroup,
+                                   @JsonProperty("feed") String feed) {
+        this.consumerProps = consumerProps;
+        for (Map.Entry<Object, Object> configItem : this.consumerProps.entrySet()) {
+            System.setProperty(configItem.getKey().toString(), configItem.getValue().toString());
```

Comment:
RocketMQ client retrieves most of its settings via System properties and all of its setting item key starts with "rocketmq.", it's unlikely to introduce any conflict.